ethereum/core/src/test/java/org/hyperledger/besu/ethereum/privacy/PrivateWorldStateReaderTest.java

```diff
@@ -0,0 +1,125 @@
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.when;
+
+import org.hyperledger.besu.ethereum.core.Account;
+import org.hyperledger.besu.ethereum.core.Address;
+import org.hyperledger.besu.ethereum.core.Hash;
+import org.hyperledger.besu.ethereum.core.WorldState;
+import org.hyperledger.besu.ethereum.worldstate.WorldStateArchive;
+
+import java.util.Optional;
+
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class PrivateWorldStateReaderTest {
+
+  private final String PRIVACY_GROUP_ID = "B1aVtMxLCUHmBVHXoZzzBgPbW/wj5axDpW9X8l91SGo=";
+  private final Bytes32 privacyGroupBytes = Bytes32.wrap(Bytes.fromBase64String(PRIVACY_GROUP_ID));
+  private final Bytes contractCode = Bytes.fromBase64String("ZXhhbXBsZQ==");
+  private final Address contractAddress = Address.ZERO;
+  private final Hash blockHash = Hash.ZERO;
+  private final Hash stateRootHash = Hash.ZERO;
+
+  @Mock private PrivateStateRootResolver privateStateRootResolver;
+  @Mock private WorldStateArchive privateWorldStateArchive;
+  @Mock private WorldState privateWorldState;
+  @Mock private Account contractAccount;
+
+  private PrivateWorldStateReader privateWorldStateReader;
+
+  @Before
+  public void before() {
+    privateWorldStateReader =
+        new PrivateWorldStateReader(privateStateRootResolver, privateWorldStateArchive);
+  }
+
+  @Test
+  public void absentPrivateWorldStateReturnsEmpty() {
+    final Bytes32 privacyGroupBytes = Bytes32.wrap(Bytes.fromBase64String(PRIVACY_GROUP_ID));
+    final Address contractAddress = Address.ZERO;
+
+    when(privateStateRootResolver.resolveLastStateRoot(eq(privacyGroupBytes), eq(blockHash)))
+        .thenReturn(stateRootHash);
+    when(privateWorldStateArchive.get(eq(stateRootHash))).thenReturn(Optional.empty());
+
+    final Optional<Bytes> maybecontractCode =
+        privateWorldStateReader.getContractCode(PRIVACY_GROUP_ID, blockHash, contractAddress);
+
+    assertThat(maybecontractCode).isNotPresent();
+  }
+
+  @Test
+  public void absentAccountReturnsEmpty() {
+    final Bytes32 privacyGroupBytes = Bytes32.wrap(Bytes.fromBase64String(PRIVACY_GROUP_ID));
+    final Address contractAddress = Address.ZERO;
+
+    when(privateStateRootResolver.resolveLastStateRoot(eq(privacyGroupBytes), eq(blockHash)))
+        .thenReturn(stateRootHash);
+    when(privateWorldStateArchive.get(eq(stateRootHash)))
+        .thenReturn(Optional.of(privateWorldState));
+    when(privateWorldState.get(eq(contractAddress))).thenReturn(null);
+
+    final Optional<Bytes> maybeContractCode =
+        privateWorldStateReader.getContractCode(PRIVACY_GROUP_ID, blockHash, contractAddress);
+
+    assertThat(maybeContractCode).isNotPresent();
```

Comment:
There are multiple cases where we do return an empty optional. I guess there is no way to distinguish between these cases, but to check the logs?

Reply:
You are right. There are different conditions that can lead to an empty result and in most cases, we can't really provide extra info to the user.